File: go/planet-time/interplanet_time/scheduling.go
package interplanet_time

// MeetingWindow represents a time window when all parties are in work hours.
type MeetingWindow struct {
	StartMs         int64
	EndMs           int64
	DurationMinutes int
}

// FindMeetingWindows scans earthDays ahead in stepMin increments and returns
// windows where both planet a and planet b are in work hours.
func FindMeetingWindows(a, b string, fromMs int64, earthDays, stepMin int) []MeetingWindow {
	if earthDays <= 0 {
		earthDays = 7
	}
	if stepMin <= 0 {
		stepMin = 15
	}
	stepMs := int64(stepMin) * 60_000
	endMs := fromMs + int64(earthDays)*EarthDayMs

	var windows []MeetingWindow
	inWindow := false
	var winStart int64

	for t := fromMs; t < endMs; t += stepMs {
		ptA := GetPlanetTime(a, t, 0)
		ptB := GetPlanetTime(b, t, 0)
		both := ptA.IsWorkHour && ptB.IsWorkHour

		if both && !inWindow {
			inWindow = true
			winStart = t
		} else if !both && inWindow {
			inWindow = false
			durMin := int((t - winStart) / 60_000)
			if durMin > 0 {
				windows = append(windows, MeetingWindow{
					StartMs:         winStart,
					EndMs:           t,
					DurationMinutes: durMin,
				})
			}
		}
	}
	if inWindow {
		durMin := int((endMs - winStart) / 60_000)
		if durMin > 0 {
			windows = append(windows, MeetingWindow{
				StartMs:         winStart,
				EndMs:           endMs,
				DurationMinutes: durMin,
			})
		}
	}
	return windows
}

File: elixir/planet-time/lib/interplanet_time/scheduling.ex
defmodule InterplanetTime.Scheduling do
  @moduledoc """
  find_meeting_windows/3 — find overlapping work windows between two planets.
  Ported verbatim from planet-time.js (Story 18.13).
  """

  alias InterplanetTime.TimeCalc
  alias InterplanetTime.Constants

  @step_ms 15 * 60_000  # 15-minute steps

  @doc """
  Find overlapping work windows between `planet_a` and `planet_b`
  over `earth_days` Earth days, starting from `start_ms` (UTC ms).

  Returns a list of maps: `%{start_ms, end_ms, duration_min}`.
  """
  def find_meeting_windows(planet_a, planet_b, opts \\ []) do
    earth_days = Keyword.get(opts, :earth_days, 7)
    start_ms   = Keyword.get(opts, :start_ms, System.os_time(:millisecond))
    end_ms     = start_ms + earth_days * Constants.earth_day_ms()

    scan(planet_a, planet_b, start_ms, end_ms, false, 0, [])
    |> Enum.reverse()
  end

  defp scan(_a, _b, t, end_ms, false, _ws, acc) when t >= end_ms, do: acc
  defp scan(_a, _b, t, end_ms, true,  ws, acc) when t >= end_ms do
    win = %{start_ms: ws, end_ms: end_ms, duration_min: div(end_ms - ws, 60_000)}
    [win | acc]
  end

  defp scan(planet_a, planet_b, t, end_ms, in_window, window_start, acc) do
    ta = TimeCalc.get_planet_time(planet_a, t)
    tb = TimeCalc.get_planet_time(planet_b, t)
    overlap = ta.is_work_hour and tb.is_work_hour

    {new_in, new_ws, new_acc} =
      cond do
        overlap and not in_window ->
          {true, t, acc}
        not overlap and in_window ->
          win = %{start_ms: window_start, end_ms: t, duration_min: div(t - window_start, 60_000)}
          {false, 0, [win | acc]}
        true ->
          {in_window, window_start, acc}
      end

    scan(planet_a, planet_b, t + @step_ms, end_ms, new_in, new_ws, new_acc)
  end
end
